### ISSUE TITLE
Difference Between Continuous Integration vs Continuous Delivery vs Continuous Deployment

### ISSUE THREAD

**Issue by abhi-debuginside** [abhi-debuginside/abhi-debuginside.github.io#5]
## Confused with continuous integration (CI), continuous delivery (CD), and continuous deployment (CD)?. You are not alone!.

In this blog, I am trying to share my understanding of the difference between CI/CD/CD. for the simplicity, we will cover only the difference between these three process.

Every DevOps experts have their own view and explanations about each method or practice.

Continous Integration
Continous integration is a development practice to ensure better quality and maintainable software development.

If a group devs are working on a software development project, each of these members will be working on the different features. After finishing an individual feature, these features will merge (or combine) with existing code base.

At this stage, there is a chance for error. By automating these integration developers can save integration time and gets an error immediately if the unit test/integration test fails.

For better output dev need to write all possible test cases to ensure a better quality of the software. There are lots of tools for CI automation.

### Continous Delivery
Continous delivery can be considered as next step of CI. After passing a build, unit test and integrated testing (CI stage), new changes are merged to pre-release branch and run a logic testing, to ensure a bug-free source code ready to release.

Continuous delivery mostly happens in the staging environment. And this is subjective based on the context of each software project.

### Continous Deployment
The continuous deployment process is all about releasing the build to production. This can be done manually or an automation workflow can do it for you. An automated release is the best way if the rollout is happening frequently.

If a new change for a release build is pushed to a production branch, this will trigger an automated deployment of the new feature to the production environment.

The key benefit of the automated continuous deployment is,

if something went wrong with a release, they can either do a hotfix and release the hotfix immediately.
Or they can even opt for roll back to the previous version or any specific version of previous releases.
Conclusion
In conclusion, the adoption of the  CI, CD, and CD depends on various variables like the type of the project, decision makers, and its targetted audience.

Combination of all the three process helps to improve the timely delivery of the better quality application to the end user.

I have come to the above conclusion after reading many blogs and talking to people who have had hands-on experience in this field.

This is completely a personal view, please feel free comment on your opinion about this post.

Thanks

References

https://en.wikipedia.org/wiki/DevOps

https://en.wikipedia.org/wiki/Continuous_integration

https://en.wikipedia.org/wiki/Continuous_delivery
